### ISSUE TITLE
Handling NaN propagation:

### ISSUE THREAD

**Issue by AidanRocke** [pauli-space/tic-tac-toe_RL#2]
It appears that there's no problem with calculating the actual loss. However, the output of the neural network is nans after the fourth sampled action. Not sure why but I can think of several culprits:

1. This might be related to the [following issue](https://github.com/tensorflow/tensorflow/issues/2540) as well as [the following problem](https://stackoverflow.com/questions/50194651/effective-method-for-accumulating-gradients-in-tensorflow).   
2. I have went from this non-differentiable log-probability:

```
def log_prob(self):
        
        log_p = self.dist.log_prob(self.action)
        
        return tf.where(tf.is_nan(log_p), tf.ones_like(log_p) * 0.1, log_p)
```

to this: 

```
def log_prob(self):
        
        ## use softmax to calculate probabilities:
        probs = tf.nn.softmax(self.policy)
        
        ## approximate the maximum probability:
        max_p = tf.pow(tf.reduce_sum(tf.pow(probs,100)),1/100)
        
        return tf.log(max_p+tf.constant(1e-8))
```

Yet, this still doesn't solve the problem so I think the issue probably lies with the first potential culprit i.e. the method for accumulating gradients.

**Comment by AidanRocke** [comment]
This problem isn't solved even when I restrict the optimisation problem to just the policy network and apply gradient clipping: 

```
self.average_loss = -1.0*tf.reduce_mean(self.reinforce_loss)
        
        #self.average_loss = -1.0*tf.reduce_mean(tf.subtract(self.reinforce_loss,self.baseline)) + \
         #                   0.5*tf.reduce_mean(tf.square(self.value_estimate-self.reward))
        
        ## collect trainable variables:
        #self.TV = tf.trainable_variables()
        
        self.TV = tf.get_collection(key = tf.GraphKeys.TRAINABLE_VARIABLES,
                                     scope= "policy_net")
        
        ## define training operations:
        self.optimizer = tf.train.AdagradOptimizer(0.01)
        
        self.accum_vars = [tf.Variable(tf.zeros_like(tv.initialized_value()), trainable=False) for tv in self.TV]                                        
        self.zero_ops = [tv.assign(tf.zeros_like(tv)) for tv in self.accum_vars]
        
        ## clip gradients:
        self.gvs_ = self.optimizer.compute_gradients(self.average_loss, self.TV)
        self.gvs = [(tf.clip_by_value(grad, -1., 1.), var) for grad, var in self.gvs_]
        #self.gvs =tf.where(tf.is_nan(self.gvs_), tf.zeros_like(self.gvs_), self.gvs_)
        
        self.accum_ops = [self.accum_vars[i].assign_add(gv[0]) for i, gv in enumerate(self.gvs)]
        
        self.train_step = self.optimizer.apply_gradients([(self.accum_vars[i], gv[1]) for i, gv in enumerate(self.gvs)])
```
 
I still get nans.

**Comment by AidanRocke** [comment]
The solution turned out to be pretty simple: https://stackoverflow.com/questions/50546197/tensorflow-removing-nans-in-accumulated-gradients/50546321#50546321

The issue was that compute_gradients returns a list of tensors.
